### PR TITLE
Marbleland Play and Leaderboards support

### PR DESCRIPTION
**Pull request by RandomityGuy**
It's all data driven.
- Added play button which launches the custom level in a new tab / launches PQ.
- Added leaderboards, fetched from various data sources.
- Also added basic malicious code check.

### CODE REVIEW

**Comment by RandomityGuy** [comment]
Added better duplicate mis detection through AST hashing.

**Comment by RandomityGuy** [comment]
Ok moved 
> Also added basic malicious code check.
> Added better duplicate mis detection through AST hashing.
to separate branches now

File: server/ts/api/api_level.ts [hunk 1]
```diff
@@ -526,6 +539,35 @@ export const initLevelApi = () => {
...
+
+		try {
+			let resp = await fetch(query.queryUrl.replace('{id}', levelId.toString()));
+			let json = await resp.json();
+
+			let respParsed = lbEndpointResponseSchema.parse(json);
+			respParsed.scores.sort((a, b) => a.placement - b.placement); // Sort the scores in place anyway, make sure we always return sorted
+			res.send(respParsed); // Just send it as it is, it's validated
+		} catch (e) {
+			res.status(400).send("400\nEndpoint schema did not return in the expected format.");
```
Comment (Vanilagy) [inline server/ts/api/api_level.ts:567]: I think this should return `502 bad gateway` here

**Review by Vanilagy — APPROVED** [review]
Left only one nitpick, the server code seems correct to me.

File: server/ts/api/api_level.ts (hunk 1 above)
Comment (RandomityGuy) [reply server/ts/api/api_level.ts:567]: Sure

**Pull request merged**